### ISSUE TITLE
`@fold` support in MSSQL

### ISSUE THREAD

**Issue by jecdohmann** [kensho-technologies/graphql-compiler#555]
@chewselene @obi1kenobi  @pmantica1 let me know your thoughts

MSSQL doesn't have an array aggregation operator so folds need to be done using alternative methods (mostly manipulating the string results of the `FOR XML PATH` operator).

**Edited to use less common delimiters and to escape nulls unambiguously**
**Edited to defer conversion of auto-escaped characters back to their original form until a post processing step SEE BELOW**

The generic structure is:

```
SELECT
    OuterVertex.property AS alias_1,
    COALESCE(STUFF(
        (SELECT 
           '~|*' + REPLACE(tbl_inner.name, '|', '||')
        FROM FoldedVertex tbl_inner
        WHERE (tbl_inner.outEdge = tbl_outer.key)
        FOR XML PATH ('')),
        1,
        3,
        ''), '!|#null!|#') AS alias_2
FROM OuterVertex AS tbl_outer
```

The XML PATH is needed to iterate over each FoldedVertex.

_x_count is captured by adding a nested subquery and the joining:

```
SELECT
    OuterVertex.property AS alias_1,
   COALESCE(STUFF(
        (SELECT 
           '~|*' + REPLACE(tbl_inner.name, '|', '||')
        FROM FoldedVertex tbl_inner
        WHERE (tbl_inner.outEdge = tbl_outer.key)
        FOR XML PATH ('')),
        1,
        3,
        ''), '!|#null!|#') AS alias_2,
    COALESCE(inner_1._x_count, 0) AS _x_count <-- here
FROM OuterVertex AS tbl_outer
LEFT JOIN ( <-- here
  SELECT 
    tbl_inner.outEdge AS group_id,
    COUNT(tbl_inner.property) AS  _x_count
  FROM FoldedVertex tbl_inner
  GROUP BY tbl_inner.group_id
) AS inner_0
ON tbl_outer.key = inner_0.outEdge
```

Applying a filter to the x_count is accomplished by putting a WHERE clause at the very outside of the query (this ensures that it is logically imposed last)

```
SELECT
    OuterVertex.property AS alias_1,
    COALESCE(STUFF(
        (SELECT 
            '~|*' + REPLACE(tbl_inner.name, '|', '||')
        FROM FoldedVertex tbl_inner
        WHERE (tbl_inner.outEdge = tbl_outer.key)
        FOR XML PATH ('')),
        1,
        3,
        ''),'!|#null!|#') AS alias_2,
    COALESCE(inner_1._x_count, 0) AS _x_count
FROM OuterVertex AS tbl_outer
LEFT JOIN (
  SELECT 
    tbl_inner.outEdge AS group_id,
    COUNT(tbl_inner.property) AS  _x_count
  FROM FoldedVertex tbl_inner
  GROUP BY tbl_inner.group_id
) AS inner_0
ON tbl_outer.key = inner_0.outEdge
WHERE inner_0._x_count >= 2 <-- here
```

Filtering on properties of the outer vertex is accomplished by embedding another SELECT at the location where the OuterVertex is usually referenced.

```
SELECT
    OuterVertex.property AS alias_1,
    COALESCE(STUFF(
      COALESCE(STUFF(
        (SELECT 
           '~|*' + REPLACE(tbl_inner.name, '|', '||')
        FROM FoldedVertex tbl_inner
        WHERE (tbl_inner.outEdge = tbl_outer.key)
        FOR XML PATH ('')),
        1,
        3,
         ''),'!|#null!|#') AS alias_2,
    COALESCE(inner_1._x_count, 0) AS _x_count
FROM (SELECT * FROM OuterVertex WHERE filtered_property > 60000) AS tbl_outer <--here
LEFT JOIN (
  SELECT 
    tbl_inner.outEdge AS group_id,
    COUNT(tbl_inner.property) AS  _x_count
  FROM FoldedVertex tbl_inner
  GROUP BY tbl_inner.group_id
) AS inner_0
ON tbl_outer.key = inner_0.outEdge
WHERE inner_0._x_count >= 2
```

Filtering on a property on one of the folded vertices is the most complicated: we need to add a filter into the select statement containing the XML Path command, and we need to introduce ANOTHER subquery to whatever COUNT subquery we may already have.

```
SELECT
    OuterVertex.property AS alias_1,
   COALESCE(STUFF(
        (SELECT 
           '~|*' + REPLACE(tbl_inner.name, '|', '||')
        FROM (SELECT * FROM FoldedVertex  WHERE filtered_property_2 = 5) tbl_inner
        WHERE (tbl_inner.outEdge = tbl_outer.key)
        FOR XML PATH ('')),
        1,
        3,
         ''), '!|#null!|#') AS alias_2,
    COALESCE(inner_1._x_count, 0) AS _x_count
FROM (SELECT * FROM OuterVertex WHERE filtered_property > 60000) AS tbl_outer 
LEFT JOIN (
  SELECT 
    tbl_inner.outEdge AS group_id,
    COUNT(tbl_inner.property) AS  _x_count
  FROM (SELECT * FROM FoldedVertex  WHERE filtered_property_2 = 5) tbl_inner <-- here
  GROUP BY tbl_inner.group_id
) AS inner_0
ON tbl_outer.key = inner_0.outEdge
WHERE inner_0._x_count >= 2
```

The above query simultaneously filters on x_count, filters on a folded vertex property and on a root vertex property.

Each definition given above is self contained and composable with all the other definitions, so we are able to express any combination of the functionalities.

All of this is pretty much the same if a second folded edge appears - just double everything (incl an extra LEFT JOIN) - and so on for third edges, etc.

Attached file contains all the queries, you can put them into this sql fiddle to run them http://sqlfiddle.com/#!18/b8803/2
[mssql_fold.txt](https://github.com/kensho-technologies/graphql-compiler/files/3602608/mssql_fold.txt)

I am not sure how to deal with post processing of getting our delimited strings back into real arrays afterwards because MSSQL doesn't support arrays

**Comment by obi1kenobi** [comment]
I can't seem to run the fiddle you posted, I'm getting the following error: `CREATE ASSEMBLY for assembly '18_d3df49.Z.Expressions.Compiler' failed because assembly '18_d3df49.Z.Expressions.Compiler' is malformed or not a pure .NET assembly. Unverifiable PE Header/native stub.`

I'm also not sure I understand your comment, especially the `they also replace & with &` bit.

For a delimiter between the concatenated field values, I'd try to pick a character that doesn't occur frequently so that we don't have to do a lot of escaping. Commas are quite frequent, and also would introduce issues if outputting floating-point numbers in some locales that use a decimal comma instead of a decimal period. I'd suggest something like `|` as a separator character. Also, make sure to consider how we'd handle concatenation of nullable fields, and how we'd distinguish `null` from `"null"` etc.

**Comment by jecdohmann** [comment]
> I can't seem to run the fiddle you posted, I'm getting the following error: `CREATE ASSEMBLY for assembly '18_d3df49.Z.Expressions.Compiler' failed because assembly '18_d3df49.Z.Expressions.Compiler' is malformed or not a pure .NET assembly. Unverifiable PE Header/native stub.`
> 
> I'm also not sure I understand your comment, especially the `they also replace & with &` bit.
> 
> For a delimiter between the concatenated field values, I'd try to pick a character that doesn't occur frequently so that we don't have to do a lot of escaping. Commas are quite frequent, and also would introduce issues if outputting floating-point numbers in some locales that use a decimal comma instead of a decimal period. I'd suggest something like `|` as a separator character. Also, make sure to consider how we'd handle concatenation of nullable fields, and how we'd distinguish `null` from `"null"` etc.

Good points. I will update the separator and the representation of null. I can represent null by wrapping it in the delimiter symbol that will ensure that if the delimiter is designed well enough to unambiguously represent a separator then it will also be strippable from the nulls. 

Also, I made a mistake with the line `they also replace & with &`. They replace & with `& amp ;` (you can't type that without the spaces in markdown without it being rendered as just &)

I think the link for the fiddle died. This fiddle isn't particularly sophisticated so I don't think it supports saving and sharing very well. Try this http://sqlfiddle.com/#!18/d3df49/6

Alternatively, the test is in the file below:
[mssql_fold_fiddle_test.txt](https://github.com/kensho-technologies/graphql-compiler/files/3616800/mssql_fold_fiddle_test.txt)

**Comment by jecdohmann** [comment]
I have edited the plan to defer all replacement of auto-escaped chars to their original form until python post processing. The reason why is that in addition to convering <,>,& to `&lt;`, etc. there is also a whole host of hex numbers which are converted to `&#0xHEX;` and it would be best to capture those with one single regex - easier done in python than in mssql.

An important design decision here is that it is possible to avoid the auto-escaping if we do `XML PATH(''), TYPE).value('.', 'VARCHAR(max)')` - the only downside is that this form isn't robust to inputs containing certain unicode characters and would need to be both preprocessed to pass through the `TYPE` function and postprocessed to undo that preprocessing. The method using just `XML PATH('')` just needs postprocessing b/c it is robust to all character inputs.

**Comment by obi1kenobi** [comment]
Closed by #613.
